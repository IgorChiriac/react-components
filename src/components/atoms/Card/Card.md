The design team has agreed on two variations of cards depending on the context:

#### Default variant

A standard or default card, which we'll assume if the type property has been omitted or set explicitly to card. As of this writing, cards of this type are non-interactable and have a border radius of 4px.

```js { "props": { "style": { "backgroundColor": "#141E64" } } }
import { Card, Text } from '@zopauk/react-components';

<Card type="card">
  <Text as="p">I'm a card ðŸ˜€ , with an explicit type prop.</Text>
</Card>;
```

#### Link variant

A card that is meant to be clickable and respond to user interactions (hover effects et al). Border radius of 8px. Note that any additional **styles and effects are meant to be added separately** by using this component as a base on which to build another with the desired effects.

Cards of type `"linkCard"` are meant to be used to build components that interact with user actions.

```js { "props": { "style": { "backgroundColor": "#141E64" } } }
import { Card, Text, Heading } from '@zopauk/react-components';

<Card type="linkCard" onClick={() => alert("You're a winner!")}>
  <Heading as="h3">New rates today! ðŸ’¸</Heading>
  <Text as="p">Find out how you can benefit the most from our new rates.</Text>
</Card>;
```
