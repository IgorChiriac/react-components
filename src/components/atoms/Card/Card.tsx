import React from 'react';
import styled from 'styled-components';
import * as colors from '../../../constants/colors';
import { maxMedia } from '../../../helpers/responsiveness';

export interface ICardProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Type of card to render
   * @default card
   */
  type?: TCardTypes;
  /**
   * @default colors.base.white
   */
  borderColor?: string;
  /**
   * @default colors.base.white
   */
  backgroundColor?: string;
  /**
   * CSS display property
   * @default block
   */
  display?: string;
}

export type TCardTypes = 'card' | 'linkCard';

const borderRadius: { [index in TCardTypes]: string } = {
  card: '4px',
  linkCard: '8px',
};

const SCard = styled.div<ICardProps>`
  background-color: ${({ backgroundColor }) => backgroundColor};
  border-color: ${({ borderColor, backgroundColor }) => borderColor || backgroundColor};
  border-radius: ${({ type }) => borderRadius[type as TCardTypes]};
  border-style: solid;
  border-width: 2px;
  display: ${(props: ICardProps) => props.display || 'block'};
  padding: 2em;

  ${({ type }) =>
    type === 'linkCard'
      ? `
    cursor: pointer;
    user-select: none;
  `
      : undefined}

  ${maxMedia.tablet`
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  `};
`;

const Card: React.FunctionComponent<ICardProps> = props => {
  return <SCard {...props} />;
};

Card.defaultProps = {
  backgroundColor: colors.base.white,
  borderColor: colors.base.white,
  type: 'card',
};

export default Card;
