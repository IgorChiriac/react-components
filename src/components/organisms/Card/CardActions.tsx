import styled from 'styled-components';

const CardActions = styled.div`
  margin-top: 24px;
  display: flex;
  justify-content: flex-end;
`;

export default CardActions;
