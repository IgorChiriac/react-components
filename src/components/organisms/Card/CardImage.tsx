import React, { ReactNode } from 'react';
import styled from 'styled-components';

type TCardImageProps = {
  url: string;
  renderWith?: ReactNode;
};

export const CardImageContainer = styled.div`
  position: relative;
  & > a,
  & > button,
  & > div {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }
`;

const Image = styled.div<Pick<TCardImageProps, 'url'>>`
  height: 100%;
  width: 100%
  background-image: url(${({ url }) => url});
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
`;

const CardImage = ({ renderWith, url }: TCardImageProps) => (
  <CardImageContainer>
    <Image url={url} />
    {renderWith}
  </CardImageContainer>
);

export default CardImage;
