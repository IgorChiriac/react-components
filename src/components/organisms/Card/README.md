### Summary

### Example

```tsx
import { Card, Button } from '@zopauk/react-components';

<Card layout="vertical" styling="secondary">
  <Card.Image url="http://placekitten.com/500/300" />
  <Card.Content>
    <Card.Heading>Silk news popular shake bring wore</Card.Heading>
    <Card.Text>
      score signal grew to studying husband way name situation flies tall at screen massage smile burst shelter harbor
      cut most program buy afraid heavy
    </Card.Text>
    <Card.Actions>
      <Button>Button</Button>
    </Card.Actions>
  </Card.Content>
</Card>;
```
