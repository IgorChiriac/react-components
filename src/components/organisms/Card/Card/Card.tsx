import styled from 'styled-components';
import { typography } from '../../../../constants/typography';
import { colors } from '../../../../constants/colors';
import { CardImageContainer } from '../CardImage/CardImage';
import CardHeading from '../CardHeading/CardHeading';
import CardText from '../CardText/CardText';

export type TCardLayout = 'horizontal' | 'vertical';
export type TCardStyling = 'primary' | 'secondary';

export type TCardProps = {
  layout?: TCardLayout;
  styling?: TCardStyling;
};

const layoutDirections: { [key in TCardLayout]: string } = {
  horizontal: 'row',
  vertical: 'column',
};

const layoutImageStyles: { [key in TCardLayout]: string } = {
  horizontal: '200px',
  vertical: '200px',
};

const headingSizes: { [key in TCardStyling]: string } = {
  secondary: typography.sizes.heading.h6,
  primary: typography.sizes.heading.h5,
};

const textSizes: { [key in TCardStyling]: string } = {
  secondary: typography.sizes.text.small,
  primary: typography.sizes.text.body,
};

const Card = styled.div<TCardProps>`
  height: 100%;
  width: 100%;
  display: flex;
  overflow: hidden;
  background-color: ${colors.white};
  border: 1px solid ${colors.greyLighter};
  border-radius: 12px;
  box-shadow: 0 1px 0 0 ${colors.greyLight};
  flex-direction: ${({ layout = 'vertical' }) => layoutDirections[layout]};
  & > ${CardImageContainer} {
    flex: 0 1 ${({ layout = 'vertical' }) => layoutImageStyles[layout]};
  }
  & ${CardHeading} {
    font-size: ${({ styling = 'secondary' }) => headingSizes[styling]};
  }
  & ${CardText} {
    font-size: ${({ styling = 'secondary' }) => textSizes[styling]};
  }
`;

export default Card;
