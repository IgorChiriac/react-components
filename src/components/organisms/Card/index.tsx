import React, { FC } from 'react';

import CardComponent, { TCardProps } from './Card/Card';
import CardActions from './CardActions/CardActions';
import CardContent from './CardContent/CardContent';
import CardImage from './CardImage';
import CardHeading from './CardHeading/CardHeading';
import CardText from './CardText';

type TCardStatic = {
  Image: typeof CardImage;
  Content: typeof CardContent;
  Actions: typeof CardActions;
  Heading: typeof CardHeading;
  Text: typeof CardText;
};

const Card: TCardStatic & FC<TCardProps> = props => <CardComponent {...props} />;

Card.Image = CardImage;
Card.Content = CardContent;
Card.Actions = CardActions;
Card.Heading = CardHeading;
Card.Text = CardText;

export default Card;
