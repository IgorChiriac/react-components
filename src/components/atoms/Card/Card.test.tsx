import { axe } from 'jest-axe';
import React from 'react';
import { render } from '@testing-library/react';
import Card from './Card';

describe('<Card type="card"/>', () => {
  it('renders default card type with explicit card type', () => {
    const { container } = render(
      <Card type="card">
        <div>Content</div>
      </Card>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('renders default card type with implicit card type with no a11y violations', async () => {
    const { container } = render(
      <Card>
        <div>Content</div>
      </Card>,
    );
    expect(container.firstChild).toMatchSnapshot();
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });
  it('renders the card type with multiple props', () => {
    const { container } = render(
      <Card type="card" borderColor="#abc" backgroundColor="#def">
        <div>Content</div>
      </Card>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<Card type="linkCard"/>', () => {
  it('renders default linkCard', () => {
    const { container } = render(<Card type="linkCard" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it('renders the component with props', () => {
    const { container } = render(
      <Card type="linkCard" borderColor="#abc" backgroundColor="#def">
        <div>Content</div>
      </Card>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
