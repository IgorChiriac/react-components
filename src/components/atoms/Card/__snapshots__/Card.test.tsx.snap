// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card type="button"/> renders the component with props 1`] = `
.c0 {
  background-color: #fff;
  border-color: #fff;
  border-radius: 1000px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="button"
>
  Click me!
</div>
`;

exports[`<Card type="button"/> renders the component with props 2`] = `
.c0 {
  background-color: #def;
  border-color: #abc;
  border-radius: 1000px;
  border-style: solid;
  border-width: 2px;
  display: inline-block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  display="inline-block"
  type="button"
>
  Click me too
</div>
`;

exports[`<Card type="card"/> renders default card type with explicit card type 1`] = `
.c0 {
  background-color: #fff;
  border-color: #fff;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="card"
>
  <div>
    Content
  </div>
</div>
`;

exports[`<Card type="card"/> renders default card type with implicit card type with no a11y violations 1`] = `
.c0 {
  background-color: #fff;
  border-color: #fff;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="card"
>
  <div>
    Content
  </div>
</div>
`;

exports[`<Card type="card"/> renders the card type with multiple props 1`] = `
.c0 {
  background-color: #def;
  border-color: #abc;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="card"
>
  <div>
    Content
  </div>
</div>
`;

exports[`<Card type="linkCard"/> renders default linkCard 1`] = `
.c0 {
  background-color: #fff;
  border-color: #fff;
  border-radius: 8px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="linkCard"
/>
`;

exports[`<Card type="linkCard"/> renders the component with props 1`] = `
.c0 {
  background-color: #def;
  border-color: #abc;
  border-radius: 8px;
  border-style: solid;
  border-width: 2px;
  display: block;
  padding: 2em;
}

@media (max-width:720px) {
  .c0 {
    border-radius: 0;
    padding-left: 1em;
    padding-right: 1em;
  }
}

<div
  class="c0"
  type="linkCard"
>
  <div>
    Content
  </div>
</div>
`;
